kernel/src/process_checking.rs

```diff
@@ -0,0 +1,356 @@
+//! Traits, types, and sample implementations of application credentials
+//! checkers, used to decide whether an application can be loaded. See
+//| the [AppID TRD](../../doc/reference/trd-appid.md).
+
+use crate::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
+use crate::hil::digest::{ClientData, ClientHash, ClientVerify};
+use crate::hil::digest::{DigestDataVerify, Sha256};
+use crate::process::{Process, ShortID, State};
+use crate::utilities::cells::OptionalCell;
+use crate::utilities::cells::TakeCell;
+use crate::utilities::leasable_buffer::{LeasableBuffer, LeasableMutableBuffer};
+use crate::ErrorCode;
+use tock_tbf::types::TbfFooterV2Credentials;
+use tock_tbf::types::TbfFooterV2CredentialsType;
+
+/// What a AppCredentialsChecker decided a particular applications credential
+/// indicates about the runnability of an application binary.
+#[derive(Debug)]
+pub enum CheckResult {
+    /// Accept the credential and run the binary.
+    Accept,
+    /// Go to the next credential or in the case of the last one fall
+    /// back to the default policy.
+    Pass,
+    /// Reject the credential and do not run the binary.
+    Reject,
+}
+
+/// Receives callbacks on whether a credential was accepted or not.
+pub trait Client<'a> {
+    fn check_done(
+        &self,
+        result: Result<CheckResult, ErrorCode>,
+        credentials: TbfFooterV2Credentials,
+        binary: &'a [u8],
+    );
+}
+
+/// Implements a Credentials Checking Policy.
+pub trait AppCredentialsChecker<'a> {
+    fn set_client(&self, client: &'a dyn Client<'a>);
+    fn require_credentials(&self) -> bool;
+
+    fn check_credentials(
+        &self,
+        credentials: TbfFooterV2Credentials,
```

Comment:
We should also be passing in the TBF header as well. That way the `Checker` can have a hard coded list of allowed TBF options and compare against them.

This follows along with the [threat model](https://github.com/tock/tock/blob/master/doc/threat_model/TBF_Headers.md) that we don't trust the headers and instead we can verify them. It also matches up with the original design discussion from https://groups.google.com/g/tock-dev/c/aduN7fHWXdI/m/P0H37in4BgAJ

I understand the headers will be authenticated with the rest of the application, but this allows locking down user loadable code. For example you could allow a customer to run applications on your SoC, so they have a private key that is accepted by the `Checker`. But then the board will ensure they don't have permissions to certain syscalls or storage.

Reply:
I agree we want this feature, but I think deciding if the kernel is OK with running an app is separate from verifying the app's credentials.

Reply:
I'm not convinced they should be separate. Part of checking the credentials is also checking that everything in the header matches what you expect.

Otherwise we need a second round of checks and more steps in the state machine which seems a bit over the top

Reply:
I don't see how we are going to avoid a second round of checks. First the system will have to check the signature or hash, and only then can it check the app itself.

Reply:
This API should not include processing arbitrary TBF options. It is about checking credentials, i.e., the authenticity and integrity of processes. It is not about arbitrary security policies; that is pushed to ShortID and the credentials themselves.

Reply:
If you pass the header in here you can avoid a second round of checks. The checker can be in charge of ensuring that the application meets all requirements.

For example a kernel could be compiled supporting two keys. If an app is signed with the "admin" key it can do anything, if it's signed with the "user" key it can only perform certain operations. The checker can then enforce this. The checker can check that the app is signed with a valid key, checks the apps signature/hash and then checks that the TBF permissions are correct.

This can all be done in a single step. So if an app doesn't pass the checker for any reason it's invalid.

@phil-levis you say that's pushed to `ShortID` but I don't see anywhere in this PR where checks are carried out against the shortID. The example above about an "admin" and "user" key also becomes more complex if we are just using a `shortID` to determine permissions instead of the keys.